Add notices and "talk to sales" when renewing plan would be a downgrade

nospame opened the pull request:
## Product Description
Adds downgrade messages, when applicable, to the "renew plan" workflow. These are the same downgrade messages shown on the "confirm plan" page when selecting a plan that will be a downgrade. Also adds the "talk to sales" button on this page when downgrade messages are shown.

![image](https://github.com/user-attachments/assets/1c7aa080-4802-46a5-bac9-291d4ba06fa3)



## Technical Summary
[SAAS-19540](https://dimagi.atlassian.net/browse/SAAS-15940)
Moves downgrade messages to their own html template partial and references this in `confirm_plan_summary.html` and `renew_plan_selection.html` Adds `downgrade_messages` to page context for the `SubscriptionRenewalView`.

Includes significant refactoring of "talk to sales" forms to facilitate adding a new version of the form for this workflow. Consolidates the lead text on these forms as per ops, it was overspecific and/or redundant.

## Safety Assurance

### Safety story
This was initially a very small change to a view to add a new item to context. The bulk of it has now become refactoring after finding the `SelectedAnnualPlanView` was almost entirely duplicative of `SelectedEnterprisePlanView`, including the `AnnualPlanContactForm`.

These views are pretty simple and the logic hasn't really been changed - they all load a "Contact Dimagi" form with a couple of pieces of information pre-populated based on what plan is currently subscribed. I've tested locally and on staging to see that each view loads correctly and the form submits as expected.

### Automated test coverage
There's a reasonable number of lines of code touched here, but the bulk of it is pretty straightforward. There's no existing automated testing for these - looking at adding it I'm not sure how much value there would be in testing each view, but I've added one test case to see that the form does make a call to send an email.

### QA Plan
Not planning QA, but willing to hear arguments otherwise.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Move downgrade_messages to html partial and conditionally include on renewal page

nospame committed:
Add downgrade messages to SubscriptionRenewalView

nospame committed:
Make current_edition a property

nospame committed:
Refactor SelectedAnnual/SelectedEnterprise plan views to use base class

nospame committed:
Refactor Enterprise/Annual plan contact forms into base class

nospame committed:
Add custom plan contact view and form

nospame committed:
Refactor SelectedPlan views to use one template

nospame committed:
Consolidate contact form lead text

nospame committed:
Add TestPlanContactForm

millerdev commented on corehq/apps/accounting/forms.py:
```diff
@@ -1993,77 +1996,39 @@ def send_message(self):
-                ),
-            )
-        )
+            kwargs['back_button_text'] = "Back to my Subscription"
+            kwargs['back_button_url'] = DomainSubscriptionView.urlname
```
I assume it's not likely to matter, but these will overwrite without warning if either are passed in. What do you think of using `setdefault` instead?
```suggestion
            kwargs.setdefault('back_button_text', "Back to my Subscription")
            kwargs.setdefault('back_button_url', DomainSubscriptionView.urlname)
```

millerdev commented on corehq/apps/accounting/forms.py:
```diff
@@ -1993,77 +1996,39 @@ def send_message(self):
+
+    def __init__(self, domain, web_user, data=None, *args, **kwargs):
+        from corehq.apps.domain.views.accounting import DomainSubscriptionView
+        kwargs['back_button_text'] = "Back to my Subscription"
+        kwargs['back_button_url'] = DomainSubscriptionView.urlname
```
Same as previous suggestion
```suggestion
        kwargs.setdefault('back_button_text', "Back to my Subscription")
        kwargs.setdefault('back_button_url', DomainSubscriptionView.urlname)
```

millerdev commented on corehq/apps/accounting/tests/test_forms.py:
```diff
@@ -404,3 +407,36 @@ def test_num_form_submitting_mobile_workers(self):
+    def setUp(self):
+        super().setUp()
+        self.domain = generator.arbitrary_domain()
+        self.web_user = generator.arbitrary_user(self.domain.name, is_webuser=True)
+
+    def tearDown(self):
+        self.domain.delete()
+        super().tearDown()
```
`self.addCleanup` is preferred to `tearDown` because it will be run even if `setUp` fails.
```suggestion
        self.addCleanup(self.domain.delete)
        self.web_user = generator.arbitrary_user(self.domain.name, is_webuser=True)
```

mjriley commented on corehq/apps/accounting/forms.py:
```diff
@@ -1944,10 +1944,13 @@ class EnterprisePlanContactForm(forms.Form):
 
     def __init__(self, domain, web_user, data=None, *args, **kwargs):
+        from corehq.apps.domain.views.accounting import SelectPlanView
+        back_button_text = kwargs.pop('back_button_text', 'Select different plan')
+        back_button_url = kwargs.pop('back_button_url', SelectPlanView.urlname)
```
What do you think about using declared parameters rather than `kwargs` here? specifically, we could change this to be:
`def __init__(self, domain, web_user, back_button_text=gettext_lazy('Select different plan'), back_button_url=None, data=None, *args, **kwargs)`, or you could use the named the args syntax like:
`def __init__(self, domain, web_user, data=None, *args, back_button_text=gettext_lazy('Select different plan'), back_button_url=None, **kwargs)`. 

I'm also curious about the need for defaults here. I assumed we provided defaults because multiple forms re-used the same values, but I notice that both `AnnualPlanContactForm` and `CustomPlanContactForm` use `back_button_text` of "Back to my Subscription", while `EnterprisePlanContactForm` relies on the default value. I wonder if it would be more straight-forward to remove the default and have each class provide its own value.

Finally, I think `back_button_url` is confusing with how it's currently used, but conveys the right idea. Currently, it's not actually a URL, but a "URL name" that is resolved later. Could we instead ask the caller to pass in a fully-formed URL? That would let future callers, who use urls that need more or less than a domain argument, to still use this class.

nospame replied:
`AnnualPlanContactForm` conditionally uses "Back to my Subscription" but falls back to the "Select different plan" default if not `on_annual_plan` -- which is to say if someone is trying to sign up for an annual plan for the first time, because they would have gotten there from the "Select Plan" page. Since there are two uses of each version, it is probably equally valid to have either of them as the default, or to have no explicit default set.

I do think the idea of passing in the url itself might make sense here. Would you do that from the form (keeping some local imports) or from the view? I don't love the idea of the view having to specify which buttons we're showing on the form, but I'm not sure how that balances against local imports.

mjriley replied:
So, I'm trying to understand what `AnnualPlanContactForm` wants to represent -- what generic concept. I think this is meant to be a form that displays `name`, `company`, and `message` fields and has configurable back and submit buttons that adhere to standardized styling. Additionally, it needs to be able to send an email with a custom title. However, I don't think this is specific to `AnnualPlanContactForm` -- the above description also describes `EnterprisePlanContactForm` and `CustomPlanContactForm`. Given that each of these forms is only used by their respective view, I think this might be a case of redundant classes. My preference would be to make the Form itself generic, which you've already done with `PlanContactForm`, while removing `EnterprisePlanContactForm`, `AnnualPlanContactForm`, and `CustomPlanContactForm`. I'd have the constructor take in properties for `back_button_text` (default to something basic like 'Back') and `back_button_url`. And I'd modify `send_message` to take in a `request_type` parameter. Then each View can use `PlanContactForm`, customizing it with the back button's text and URL, and providing its own request text when calling `send_message`. 

That said, I mentioned these were non-blocking suggestions. I'll defer to your best judgment on what you think makes the most sense.

nospame replied:
I waffled on this myself a bit (how much refactoring is _too_ much refactoring), but ultimately I think you're right on the class duplication. The subclasses aren't really doing anything here when the form is truly identical, sending 99% the same message to the same place.

mjriley commented on corehq/apps/accounting/forms.py:
```diff
@@ -1993,77 +1996,39 @@ def send_message(self):
-        self.helper.label_class = 'col-sm-3 col-md-2'
-        self.helper.field_class = 'col-sm-9 col-md-8 col-lg-6'
-        self.helper.form_class = "form-horizontal"
+        from corehq.apps.domain.views.accounting import DomainSubscriptionView
```
Seeing the local import was a smell for me. This would be a reason to allow the caller to pass in the entire URL, because then this Form doesn't need to be aware of anything from the View.

mjriley commented on corehq/apps/accounting/templates/accounting/partials/confirm_plan_summary.html:
```diff
@@ -104,18 +104,5 @@ <h4>
-      </ul>
-    </div>
-  {% endfor %}
+  {% include 'accounting/partials/downgrade_messages.html' %}
```
Can we use either the ["include with" or "include with only"](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#include) syntax here? Historically, I've found it hard to trace through multiple template layers to find what defines a variable. In this case, _downgrade_messages.html_ is tightly defined to only care about `downgrade_messages`, so I think explicitly passing that variable to the template here would make it easier to reason about

nospame replied:
I hadn't seen that syntax before, but I think it makes sense here for sure. Thanks!

mjriley commented on corehq/apps/domain/views/accounting.py:
```diff
@@ -1118,18 +1119,19 @@ def can_change_subscription(self):
     step_title = gettext_lazy("Contact Dimagi")
-    edition = SoftwarePlanEdition.ENTERPRISE
+    contact_form = None
+    lead_text = _("Please submit the following form and a member of our sales team will be "
+                 "in touch shortly about your plan.")
```
Because class variables get initialized immediately, `lead_text` needs to use `gettext_lazy` (or it will always have the translation from the initial execution context).

mjriley commented on corehq/apps/domain/views/accounting.py:
```diff
@@ -1549,49 +1548,64 @@ class SubscriptionRenewalView(PlanViewBase, SubscriptionMixin):
         return {'monthly_plan': monthly_plan.user_facing_description,
                 'annual_plan': annual_plan.user_facing_description}
 
+    @property
+    def downgrade_messages(self):
+        downgrades = get_change_status(self.subscription.plan_version, self.monthly_plan_version)[1]
+        downgrade_handler = DomainDowngradeStatusHandler(
+            self.domain_object, self.monthly_plan_version, downgrades,
+        )
+        return downgrade_handler.get_response()
```
Can we make this a method rather than a property? I like when properties adhere to property-like expectations -- i.e., lightweight, quick runtime. Specifically with the `downgrade_handler.get_response()` line, it looks like, in order for it to resolve something like the `CLOUDCARE` privileges, it needs to pull all a domain's application documents from Couch, which could be quite expensive.

One specific reason I've run into to avoid complicated properties is that the Visual Studio Code and PyCharm debuggers immediately evaluate properties they run into during breakpoints. That can cause unexpected behaviors or just general slowness.

mjriley reviewed (changes requested):
The overall logic looks good, but there are some translation issues that should be fixed. Left some comments about non-blocking structural changes

nospame committed:
PR feedback: pass in button w/ fully formed url

nospame committed:
Additional PR feedback:

- use request_type instead of subject_tag
- downgrade_messages as a method
- pass only downgrade_messages to html template
- gettext_lazy for lead_text
- update test

mjriley reviewed (approved):
Appreciate the changes!

nospame committed:
Remove 99% identical PlanContactForm subclasses

nospame committed:
Further refactor contact form views

- define contact_form on base class with variable back_button
